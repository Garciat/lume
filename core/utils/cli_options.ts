import { parseArgs } from "../../deps/cli.ts";
import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port"],
    boolean: ["serve", "open"],
    alias: { serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  const serveMode = cli.serve || cli._[0] === "cms";

  // Detect location and port
  let port: number;
  let location: URL;

  if (serveMode) {
    location = cli.location
      ? new URL(cli.location)
      : new URL("http://localhost");

    if (cli.port) {
      port = parseInt(cli.port);
      location.port = port.toString();
    } else if (location.port) {
      port = parseInt(location.port);
    } else if (options.server?.port) {
      port = options.server.port;
      location.port = port.toString();
    } else {
      port = location.protocol === "https:" ? 443 : 3000;
      location.port = port.toString();
    }
  } else {
    location = cli.location
      ? new URL(cli.location)
      : (options.location as URL | undefined) || new URL("http://localhost");

    if (cli.port) {
      port = parseInt(cli.port);
      location.port = port.toString();
    } else if (location.port) {
      port = parseInt(location.port);
    } else {
      port = location.protocol === "https:" ? 443 : 80;
    }
  }

  options.location = location;
  options.server ||= {};
  options.server.port = port;

  if (cli.open) {
    options.server.open = cli.open;
  }

  return options;
}
