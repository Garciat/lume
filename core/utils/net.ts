export function localIp(): string | undefined {
  // Try/catch for https://github.com/denoland/deno/issues/25420
  try {
    for (const info of Deno.networkInterfaces()) {
      if (info.family !== "IPv4" || info.address.startsWith("127.")) {
        continue;
      }

      return info.address;
    }
  } catch {
    return undefined;
  }
}
