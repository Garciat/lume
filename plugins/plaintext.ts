import { plainText } from "../deps/remove-markdown.ts";

import type { RemoveMarkdownOptions } from "../deps/remove-markdown.ts";
import type Site from "../core/site.ts";

export interface Options extends RemoveMarkdownOptions {
}

export const defaults: Options = {
  stripListLeaders: true,
  gfm: true,
  useImgAltText: true,
  replaceLinksWithURL: false,
};

/**
 * Plugin to convert a markdown or HTML string to plain text
 * @see https://lume.land/plugins/plaintext/
 */
export function plaintext(userOptions?: Options) {
  const options = { ...defaults, ...userOptions };

  return (site: Site) => {
    site.filter("plaintext", (text?: unknown, opts?: Options) => {
      if (typeof text === "string") {
        return plainText(text, opts || options);
      }

      return text;
    });
  };
}

export default plaintext;

/** Extends Helpers interface */
declare global {
  namespace Lume {
    export interface Helpers {
      /** @see https://lume.land/plugins/plaintext/ */
      plaintext: <T>(text?: T, options?: Options) => T;
    }
  }
}
