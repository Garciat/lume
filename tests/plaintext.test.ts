import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import plaintext from "../plugins/plaintext.ts";

Deno.test("Plain text filter", async (t) => {
  const site = getSite({
    src: "plaintext",
  });

  site.use(plaintext());

  await build(site);
  await assertSiteSnapshot(t, site);
});
