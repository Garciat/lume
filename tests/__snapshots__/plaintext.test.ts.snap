export const snapshot = {};

snapshot[`Plain text filter 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
  ],
  src: [
    "/",
    "/index.vto",
  ],
}
`;

snapshot[`Plain text filter 2`] = `[]`;

snapshot[`Plain text filter 3`] = `
[
  {
    content: '<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Hello world</title>
  <meta name="description" content="This is a description in multiple lines With list elements and links">
</head>
<body>
  <h1>Hello <strong>world</strong></h1>

  <p>This is a   description<br>
in multiple lines</p>
<ul>
<li>With</li>
<li>list</li>
<li>elements</li>
<li><a href="#url">and     links</a></li>
</ul>
</body>
</html>
',
    data: {
      basename: "index",
      children: '<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Hello world</title>
  <meta name="description" content="This is a description in multiple lines With list elements and links">
</head>
<body>
  <h1>Hello <strong>world</strong></h1>

  <p>This is a   description<br>
in multiple lines</p>
<ul>
<li>With</li>
<li>list</li>
<li>elements</li>
<li><a href="#url">and     links</a></li>
</ul>
</body>
</html>
',
      content: '<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>{{ title |> plaintext }}</title>
  <meta name="description" content="{{ description |> plaintext }}">
</head>
<body>
  <h1>{{ title |> md(true) }}</h1>

  {{ description |> md }}
</body>
</html>
',
      date: [],
      description: "This is a   description<br>
in multiple lines

- With
- list
- elements
- [and     links](#url)
",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Hello **world**",
      url: "/",
    },
    src: {
      asset: false,
      ext: ".vto",
      path: "/index",
      remote: undefined,
    },
  },
]
`;
